File: windows-agent/LockedAgent/ExamSession.cs
using System.IO;
using System.Net.Http;
using System.Windows;
using System.Windows.Threading;

namespace LockedAgent;

/// <summary>
/// Orchestrates one exam session end to end. While the room is "waiting",
/// Excel is open in the background, unlocked, with no surveillance armed. The
/// moment the teacher starts the room — or immediately, if this student joins
/// after it has already started — every surveillance system arms at once:
/// fullscreen lockdown, focus watcher, keyboard hook, and periodic
/// screenshots. When the teacher ends the room, everything disarms and the
/// workbook is saved and uploaded automatically.
/// </summary>
public sealed class ExamSession
{
    private readonly string _baseUrl;
    private readonly JoinResponse _join;
    private readonly HttpClient _http = new();

    private BackendClient? _backend;
    private ExcelLauncher? _launcher;
    private FocusWatcher? _focusWatcher;
    private KeyboardHook? _keyboardHook;
    private ScreenshotService? _screenshotService;
    private OverlayWindow? _overlay;
    private WaitingWindow? _waitingWindow;
    private TestExitWindow? _testExitWindow;
    private DispatcherTimer? _heartbeatTimer;
    private bool _locked;

    public ExamSession(string baseUrl, JoinResponse join)
    {
        _baseUrl = baseUrl;
        _join = join;
    }

    public async Task StartAsync()
    {
        _backend = new BackendClient(_http, _baseUrl, _join.SessionId, _join.Lifecycle);
        _backend.Excluded += () => Application.Current.Dispatcher.Invoke(Exclude);
        _backend.RoomStarted += () => Application.Current.Dispatcher.Invoke(LockDown);
        _backend.RoomEnded += () => Application.Current.Dispatcher.Invoke(() => _ = EndExamAsync());
        _backend.StartPolling();

        string? examFilePath = null;
        if (_join.ExamFileAvailable)
        {
            var bytes = await _backend.DownloadExamFileAsync();
            if (bytes is not null)
            {
                examFilePath = Path.Combine(Path.GetTempPath(), $"locked-exam-{Guid.NewGuid():N}.xlsx");
                await File.WriteAllBytesAsync(examFilePath, bytes);
            }
        }

        _launcher = new ExcelLauncher();
        _launcher.Launch(examFilePath);

        _heartbeatTimer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(15) };
        _heartbeatTimer.Tick += async (_, _) => await _backend.SendHeartbeatAsync();
        _heartbeatTimer.Start();

        if (_join.Lifecycle == "started")
        {
            LockDown();
        }
        else
        {
            _waitingWindow = new WaitingWindow(_join.Config.ExamTitle);
            _waitingWindow.Show();
        }
    }

    /// <summary>Arms every surveillance system at once: fullscreen, focus
    /// watcher, keyboard hook, screenshots. Idempotent — a late joiner calls
    /// this from StartAsync directly, everyone else from the RoomStarted
    /// event, never both.</summary>
    private void LockDown()
    {
        if (_locked || _launcher is null) return;
        _locked = true;

        _waitingWindow?.Close();
        _waitingWindow = null;

        _launcher.EnterFullscreenLockdown();

        _focusWatcher = new FocusWatcher(_launcher.ProcessId);
        _focusWatcher.FocusLost += OnFocusLost;
        _focusWatcher.FocusRestored += OnFocusRestored;

        _keyboardHook = new KeyboardHook();
        _keyboardHook.Install();

        _screenshotService = new ScreenshotService(
            _http, _baseUrl, _join.SessionId, _join.Config.Screenshot);
        _screenshotService.Start();

        if (_join.IsTest)
        {
            _testExitWindow = new TestExitWindow();
            _testExitWindow.ExitRequested += () => Application.Current.Dispatcher.Invoke(ExitTestMode);
            _testExitWindow.Show();
        }
    }

    /// <summary>Tester clicked "Quitter le test": disarm everything exactly
    /// like a real end-of-exam, but record a distinct event so this never
    /// shows up as a real exclusion on the teacher's dashboard.</summary>
    private void ExitTestMode()
    {
        Disarm();
        _ = _backend!.SendEventAsync("test_exit");
        _launcher?.Dispose();
        _ = _backend?.DisposeAsync();
        Application.Current.Shutdown();
    }

    private void OnFocusLost()
    {
        Application.Current.Dispatcher.Invoke(() =>
        {
            _overlay = new OverlayWindow(_join.Config.CountdownSeconds);
            _overlay.CountdownExpired += OnCountdownExpired;
            _overlay.Show();
            _overlay.StartCountdown();
        });
        _ = _backend!.SendEventAsync("focus_lost");
    }

    private void OnFocusRestored()
    {
        Application.Current.Dispatcher.Invoke(() =>
        {
            _overlay?.CancelCountdown();
            _overlay = null;
        });
        _ = _backend!.SendEventAsync("focus_returned");
    }

    private void OnCountdownExpired()
    {
        _ = _backend!.SendEventAsync("excluded", new { reason = "countdown_expired" });
        Exclude();
    }

    /// <summary>Teacher clicked "Terminer l'examen": disarm everything, save
    /// the workbook to a path we control, and upload it — see
    /// api/submission.ts on the backend.</summary>
    private async Task EndExamAsync()
    {
        Disarm();
        try
        {
            if (_launcher is not null)
            {
                var bytes = _launcher.SaveAndReadBytes();
                await _backend!.UploadSubmissionAsync(bytes, $"{_join.Config.ExamTitle}.xlsx");
            }
        }
        catch
        {
            // Best-effort: the teacher's dashboard shows whichever submissions
            // made it through; a save/upload failure shouldn't hang the agent.
        }
        finally
        {
            _launcher?.Dispose();
            _ = _backend?.DisposeAsync();
            Application.Current.Shutdown();
        }
    }

    private void Exclude()
    {
        Disarm();
        _launcher?.Dispose();
        _ = _backend?.DisposeAsync(); // app is shutting down right after; fire-and-forget is fine
        // TODO: show a dedicated "vous avez été exclu" screen and terminate Excel
        // gracefully before shutting down, instead of exiting the whole agent.
        Application.Current.Shutdown();
    }

    private void Disarm()
    {
        _keyboardHook?.Dispose();
        _focusWatcher?.Dispose();
        _screenshotService?.Dispose();
        _heartbeatTimer?.Stop();
        _overlay?.CancelCountdown();
        _waitingWindow?.Close();
        _testExitWindow?.Close();
        _testExitWindow = null;
    }
}

File: mac-agent/Sources/LockedAgentMac/ExamSession.swift
import Cocoa

/// Orchestrates one exam session end to end. While the room is "waiting",
/// Excel is open in the background, unlocked, with no surveillance armed.
/// The moment the teacher starts the room — or immediately, if this student
/// joins after it has already started — every surveillance system arms at
/// once: fullscreen lockdown, focus watcher, keyboard blocker, and periodic
/// screenshots. When the teacher ends the room, everything disarms and the
/// workbook is saved and uploaded automatically. Mirrors the Windows agent's
/// ExamSession.cs structure and method names one-for-one.
final class ExamSession {
    private let baseUrl: String
    private let join: JoinResponse
    private let authToken: String
    private let backend: BackendClient
    private let excel = ExcelController()

    private var focusWatcher: FocusWatcher?
    private var keyboardBlocker: KeyboardBlocker?
    private var screenshotService: ScreenshotService?
    private var overlay: OverlayWindowController?
    private var waitingWindow: WaitingWindowController?
    private var testExitWindow: TestExitWindowController?
    private var heartbeatTask: Task<Void, Never>?
    private var locked = false

    init(baseUrl: String, join: JoinResponse, authToken: String) {
        self.baseUrl = baseUrl
        self.join = join
        self.authToken = authToken
        self.backend = BackendClient(
            baseURL: baseUrl, sessionId: join.sessionId, initialLifecycle: join.lifecycle, authToken: authToken)
    }

    func start() async {
        backend.onExcluded = { [weak self] in
            DispatchQueue.main.async { self?.exclude() }
        }
        backend.onRoomStarted = { [weak self] in
            DispatchQueue.main.async { self?.lockDown() }
        }
        backend.onRoomEnded = { [weak self] in
            Task { await self?.endExam() }
        }
        backend.startPolling()

        var examFilePath: String?
        if join.examFileAvailable, let data = await backend.downloadExamFile() {
            let path = NSTemporaryDirectory() + "locked-exam-\(UUID().uuidString).xlsx"
            FileManager.default.createFile(atPath: path, contents: data)
            examFilePath = path
        }

        excel.launch(filePath: examFilePath)

        heartbeatTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 15_000_000_000)
                if Task.isCancelled { break }
                await backend.sendHeartbeat()
            }
        }

        if join.lifecycle == "started" {
            lockDown()
        } else {
            let waiting = WaitingWindowController(examTitle: join.config.examTitle)
            waiting.showWindow(nil)
            waitingWindow = waiting
        }
    }

    /// Arms every surveillance system at once: fullscreen, focus watcher,
    /// keyboard blocker, screenshots. Idempotent — a late joiner calls this
    /// from start() directly, everyone else from onRoomStarted, never both.
    private func lockDown() {
        guard !locked else { return }
        locked = true

        waitingWindow?.window?.close()
        waitingWindow = nil

        excel.enterFullscreenLockdown()

        let watcher = FocusWatcher()
        watcher.onFocusLost = { [weak self] in self?.onFocusLost() }
        watcher.onFocusRestored = { [weak self] in self?.onFocusRestored() }
        watcher.install()
        focusWatcher = watcher

        let blocker = KeyboardBlocker()
        blocker.install()
        keyboardBlocker = blocker

        let perMinute = join.config.screenshot.enabled ? join.config.screenshot.perMinute : 0
        let screenshots = ScreenshotService(
            baseUrl: baseUrl, sessionId: join.sessionId, authToken: authToken, perMinute: perMinute)
        screenshots.start()
        screenshotService = screenshots

        if join.isTest {
            let testExit = TestExitWindowController()
            testExit.onExitRequested = { [weak self] in self?.exitTestMode() }
            testExit.showWindow(nil)
            testExitWindow = testExit
        }
    }

    private func onFocusLost() {
        let overlayController = OverlayWindowController(countdownSeconds: join.config.countdownSeconds)
        overlayController.onCountdownExpired = { [weak self] in self?.onCountdownExpired() }
        overlayController.showWindow(nil)
        overlayController.startCountdown()
        overlay = overlayController
        Task { await backend.sendEvent(type: "focus_lost") }
    }

    private func onFocusRestored() {
        overlay?.cancelCountdown()
        overlay = nil
        Task { await backend.sendEvent(type: "focus_returned") }
    }

    private func onCountdownExpired() {
        Task { await backend.sendEvent(type: "excluded", payload: ["reason": "countdown_expired"]) }
        exclude()
    }

    /// Teacher clicked "Terminer l'examen": disarm everything, save the
    /// workbook to a path we control, and upload it.
    private func endExam() async {
        disarm()
        if let bytes = excel.saveAndReadBytes() {
            await backend.uploadSubmission(fileData: bytes, filename: "\(join.config.examTitle).xlsx")
        }
        excel.quit()
        backend.stopPolling()
        NSApp.terminate(nil)
    }

    private func exclude() {
        disarm()
        excel.quit()
        backend.stopPolling()
        // TODO: show a dedicated "vous avez été exclu" screen before quitting,
        // instead of terminating the whole agent outright.
        NSApp.terminate(nil)
    }

    /// Tester clicked "Quitter le test": disarm everything exactly like a
    /// real end-of-exam, but record a distinct event so this never shows up
    /// as a real exclusion on the teacher's dashboard.
    private func exitTestMode() {
        disarm()
        Task { await backend.sendEvent(type: "test_exit") }
        excel.quit()
        backend.stopPolling()
        NSApp.terminate(nil)
    }

    private func disarm() {
        keyboardBlocker?.uninstall()
        focusWatcher?.uninstall()
        screenshotService?.stop()
        heartbeatTask?.cancel()
        overlay?.cancelCountdown()
        waitingWindow?.window?.close()
        testExitWindow?.window?.close()
        testExitWindow = nil
    }
}
